File: ruby/balancing_parentheses.rb
def balancing_parentheses(string)
  # type your code in here

  split_string = string.split('')

  if split_string.length < 1 
    return split_string.length
  end

  opening = split_string.index("(")
  if opening 
    following = split_string.slice((opening + 1)..(split_string.length))
    closing = following.index(")")
    if closing
      beginning = []
      if opening > 0
        beginning = split_string.slice(0..(opening - 1))
      end
      ending = []
      if following.length > 1
        if closing == 0
          ending = following.slice((closing + 1)..following.length)
        elsif closing > 0
          ending = following.slice(0..(closing - 1)) + following.slice((closing + 1)..following.length)
        end
      end
      new_array = beginning + ending
      new_string = new_array.join('')
      balancing_parentheses(new_string)
    else
      split_string.length
    end
  else
    split_string.length
  end

end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts balancing_parentheses('(()())')

  puts "Expecting: 3"
  puts balancing_parentheses('()()())))')

  puts "Expecting: 2"
  puts balancing_parentheses('()))')

  puts "Expecting: 5"
  puts balancing_parentheses('(((((')

  puts "Expecting: 1"
  puts balancing_parentheses(')')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/balancing_parentheses.js
function balancingParentheses(string) {

  if(string.length == 0){
    return string.length
  }

  let splitString = string.split('')

  let opening = splitString.indexOf("(")
  if(opening >= 0){
    let following = splitString.slice(opening + 1)
    let closing = following.indexOf(")")
    if(closing >= 0){
      let beginning = splitString.slice(0, opening)
      let ending = [...following.slice(0, closing), ...following.slice(closing + 1)]
      let newArray = [...beginning, ...ending]
      let newString = newArray.join('')
      return balancingParentheses(newString)
    } else {
      return splitString.length
    }
  } else {
    return splitString.length
  }

}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log(balancingParentheses('(()())'));

  console.log("Expecting: 3");
  console.log(balancingParentheses('()()())))'));

  console.log("Expecting: 2");
  console.log(balancingParentheses('()))'));

  console.log("Expecting: 5");
  console.log(balancingParentheses('((((('));

  console.log("Expecting: 1");
  console.log(balancingParentheses(')'));
}

module.exports = balancingParentheses;

// Please add your pseudocode to this file
  // split string into array
  // find first instance of "("
    // if it exists, search remainder of array (everything after the "(") to find first instance of ")"
      // if both exist, remove them from array, turn array back into string, and rerun function
    // if closing parenthesis does not exist, return length of array
  // if there are no more opening parentheses, return length of array


// And a written explanation of your solution
